private/bufpkg/bufplugin/bufplugindocker/docker.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufplugindocker
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"net"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufplugin/bufpluginconfig"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stringid"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type Client interface {
+	// Build creates a Docker image for the plugin using the Dockerfile.plugin and plugin config.
+	Build(ctx context.Context, dockerfile io.Reader, config *bufpluginconfig.Config, options ...BuildOption) (*BuildResponse, error)
+	// Push the Docker image to the remote registry.
+	Push(ctx context.Context, image string, auth RegistryAuth) (*PushResponse, error)
+	// Delete removes the Docker image from local Docker Engine.
+	Delete(ctx context.Context, image string) (*DeleteResponse, error)
+	// Close releases any resources used by the underlying Docker client.
+	Close() error
+}
+
+type BuildOption func(*buildParams)
```

Comment:
The un-exported struct should match the exported name.
```suggestion
type BuildOption func(*buildOptions)
```